core/services/relay/evm/functions/logpoller_wrapper.go

```diff
@@ -2,43 +2,98 @@ package functions
 
 import (
 	"context"
+	"encoding/hex"
 	"sync"
+	"time"
 
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
 	"github.com/ethereum/go-ethereum/common"
 
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/functions/generated/functions_coordinator"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/functions/generated/functions_router"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/functions/config"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
 	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
 	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
 	evmRelayTypes "github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
 )
 
 type logPollerWrapper struct {
-	routerContract  common.Address
-	contractVersion uint32
-	logPoller       logpoller.LogPoller
-	subscribers     map[string]evmRelayTypes.RouteUpdateSubscriber
-	mu              sync.RWMutex
-	lggr            logger.Logger
+	utils.StartStopOnce
+	contractAddress     common.Address
+	client              client.Client
+	pluginConfig        config.PluginConfig
+	logPoller           logpoller.LogPoller
+	subscribers         map[string]evmRelayTypes.RouteUpdateSubscriber
+	mu                  sync.RWMutex
+	lggr                logger.Logger
+	shutdownWaitGroup   sync.WaitGroup
+	serviceContext      context.Context
+	serviceCancel       context.CancelFunc
+	chStop              chan struct{}
+	previousCoordinator common.Address
+	activeCoordinator   common.Address
+	proposedCoordinator common.Address
 }
 
 var _ evmRelayTypes.LogPollerWrapper = &logPollerWrapper{}
 
-func NewLogPollerWrapper(routerContract common.Address, contractVersion uint32, logPoller logpoller.LogPoller, lggr logger.Logger) evmRelayTypes.LogPollerWrapper {
+func NewLogPollerWrapper(contractAddress common.Address, client client.Client, pluginConfig config.PluginConfig, logPoller logpoller.LogPoller, lggr logger.Logger) evmRelayTypes.LogPollerWrapper {
 	return &logPollerWrapper{
-		routerContract:  routerContract,
-		contractVersion: contractVersion,
+		contractAddress: contractAddress,
+		client:          client,
+		pluginConfig:    pluginConfig,
 		logPoller:       logPoller,
 		subscribers:     make(map[string]evmRelayTypes.RouteUpdateSubscriber),
 		lggr:            lggr,
+		chStop:          make(chan struct{}),
 	}
 }
 
 // TODO(FUN-381): Implement LogPollerWrapper with an API suiting all users.
 func (l *logPollerWrapper) Start(context.Context) error {
-	// TODO periodically update active and proposed coordinator contract addresses
-	// update own Logpoller filters accordingly
-	// push to all subscribers
-	return nil
+	return l.StartOnce("LogPollerWrapper", func() error {
+		l.serviceContext, l.serviceCancel = context.WithCancel(context.Background())
+		l.shutdownWaitGroup.Add(4)
```

Comment:
This should match the number of goroutines you're launching. In this case it's only 1 and only if we're v1. For v0 we don't need any goroutines because the contract address is static.